### PR TITLE
Generate temporary key pairs in AWS

### PR DESCRIPTION
**Pull request by orien**
It can be rather cumbersome to configure AWS key pairs. You have to add them in AWS and then ensure they're installed on CI servers and/or available to developers.

Instead of all that manual work, let's get AMI Spec to do it for us!

I propose that if a key name isn't provided via the command line, then a temporary key pair is generated in AWS and used for the duration of the testing. Afterward a best effort is given to delete the key pair.

### CODE REVIEW

File: lib/ami_spec/aws_key_pair.rb [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+      @key_name = "#{key_name_prefix}#{SecureRandom.uuid}"
+      @logger = logger
+    end
+
+    attr_reader :key_name, :key_file
+
+    def create
+      @logger.info "Creating temporary AWS key pair: #{@key_name}"
+      @key_pair = @ec2.create_key_pair(key_name: @key_name)
+      @temp_file = Tempfile.new('key')
```
Comment (orien) [inline lib/ami_spec/aws_key_pair.rb]: By using a `Tempfile` object, the private key will be deleted from the filesystem when the Ruby process terminates.

File: README.md [hunk 2]
```diff
@@ -37,20 +37,23 @@ Options:
   -s, --specs=<s>                       The directory to find ServerSpecs
   -u, --subnet-id=<s>                   The subnet to start the instance in
   -k, --key-name=<s>                    The SSH key name to assign to instances
+  -k, --key-name=<s>                    The SSH key name to assign to instances. If not provided a
```
Comment (patrobinson) [inline README.md]: Should the line above be removed?

File: lib/ami_spec/aws_key_pair.rb [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+      @temp_file.write(@key_pair.key_material)
+      @temp_file.close
+      @key_file = Pathname.new(@temp_file.path)
+    end
+
+    def delete
+      @logger.info "Deleting temporary AWS key pair: #{@key_name}"
+      @key_pair.delete
```
Comment (patrobinson) [inline lib/ami_spec/aws_key_pair.rb]: Does this delete it in AWS?

File: README.md (hunk 2 above)
Comment (orien) [reply README.md]: Whoops, yes it should.

File: lib/ami_spec/aws_key_pair.rb (hunk 3 above)
Comment (orien) [reply lib/ami_spec/aws_key_pair.rb]: Yes, exactly. It gets delegated to the client ([code](https://github.com/aws/aws-sdk-ruby/blob/master/gems/aws-sdk-ec2/lib/aws-sdk-ec2/key_pair.rb#L173-L189)).

**Pull request merged**